### PR TITLE
Kubernetes Secret Kubeconfig  integration

### LINKED ISSUE #13: server reference to a crossplane k8s cluster
Is there anyway to add a reference to a k8s cluster that is controlled through crossplane?

### PR DESCRIPTION
**Pull request by stevendborrelli**
### Description of your changes

Ready for review. 

- Add ability to use a `kubeconfig` secret generated by Crossplane. 
- Converts ClusterParameters `Name`, `Server`, and `Config.TLSClientConfig` to pointers as they are optional.

Fixes #13 

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

- I ran this for several hours against a cluster with some debugging enabled to monitor the secret versions and ensure every time the referenced secret was updated our controller requested an update. 
- Deployed an application against the cluster.

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by janwillies** [comment]
Thanks for taking this on @stevendborrelli! This is our most wanted feature request. Is there anything you want to discuss before doing further work here?

**Comment by stevendborrelli** [comment]
I am still doing some testing and refactoring. 

There are a couple of things I'd like some clarity on: 
1.  https://github.com/crossplane-contrib/provider-argocd/blob/14aa0b4a69a50094f423321706e25aa4baa69af7/pkg/controller/cluster/controller.go#L118

is causing cluster deletes to be skipped: 

```go
	if meta.WasDeleted(cr) && meta.GetExternalName(cr) == observedCluster.Name {
		// ArgoCD Cluster resource ignores the name field. This detects the deletion of the default cluster resource.
		return managed.ExternalObservation{}, nil
	}
```

**Comment by stevendborrelli** [comment]
To test this PR, you can use the EKS example for the AWS official providers at : <https://github.com/upbound/provider-aws/blob/main/examples/eks/cluster.yaml>. 

The secret will be written to `cluster-conn` in `upbound-system`.  

The example in `examples/cluster/cluster-kubeconfig.yaml` will use this secret to register the cluster to ArgoCD.

File: apis/cluster/v1alpha1/types.go [hunk 1]
```diff
@@ -24,9 +24,11 @@ import (
 
 // ClusterParameters define the desired state of an ArgoCD Cluster
 type ClusterParameters struct {
-	// Server is the API server URL of the Kubernetes cluster
+	// Server is the API server URL of the Kubernetes cluster. Optional if using a kubeconfig
+	// +optional
 	Server string `json:"server"`
```
Comment (MisterMX) [inline apis/cluster/v1alpha1/types.go:29]: Optional fields should be declared as pointers.

File: apis/cluster/v1alpha1/types.go [hunk 2]
```diff
@@ -24,9 +24,11 @@ import (
...
-	// Name of the cluster. If omitted, will use the server address
+	// Name of the cluster. If omitted, will use the server address. Optional if using a kubeconfig
+	// +optional
 	Name string `json:"name"`
```
Comment (MisterMX) [inline apis/cluster/v1alpha1/types.go:32]: Optional fields should be declared as pointers.

File: apis/cluster/v1alpha1/types.go [hunk 3]
```diff
@@ -59,13 +61,17 @@ type ClusterConfig struct {
 	// +optional
 	BearerTokenSecretRef *SecretReference `json:"bearerTokenSecretRef,omitempty"`
 	// TLSClientConfig contains settings to enable transport layer security
+	// +optional
 	TLSClientConfig `json:"tlsClientConfig"`
```
Comment (MisterMX) [inline apis/cluster/v1alpha1/types.go:65]: Optional fields should be declared as pointers.

File: examples/cluster/cluster-kubeconfig.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+kind: Cluster
+metadata:
+  name: example-cluster-kubeconfig
+spec:
+  forProvider:
+    name: example-cluster-kubeconfig
+    config:
+      kubeconfigSecretRef:
+        name: cluster-conn 
+        namespace: upbound-system
```
Comment (MisterMX) [inline examples/cluster/cluster-kubeconfig.yaml:12]:
```suggestion
        namespace: crossplane-system
```

File: pkg/controller/cluster/controller.go [hunk 5]
```diff
@@ -117,7 +120,9 @@ func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.Ex
 	}
 	if meta.WasDeleted(cr) && meta.GetExternalName(cr) == observedCluster.Name {
 		// ArgoCD Cluster resource ignores the name field. This detects the deletion of the default cluster resource.
-		return managed.ExternalObservation{}, nil
+		return managed.ExternalObservation{
+			ResourceExists: true,
+		}, nil
```
Comment (MisterMX) [inline pkg/controller/cluster/controller.go:125]: This change should be part of a separate PR.

File: pkg/controller/cluster/controller.go [hunk 6]
```diff
@@ -406,6 +427,18 @@ func isEqualExecProviderConfig(p *v1alpha1.ExecProviderConfig, r *argocdv1alpha1
 	return true
 }
 
+func isEqualFromKubeconfig(p *v1alpha1.ClusterParameters, r *argocdv1alpha1.Cluster) bool {
```
Comment (MisterMX) [inline pkg/controller/cluster/controller.go:430]: This function does not seem to be used anywhere.

File: pkg/controller/cluster/controller.go [hunk 7]
```diff
@@ -473,3 +516,52 @@ func (e *external) getPayload(ctx context.Context, ref *v1alpha1.SecretReference
...
+		CAData:     restConfig.CAData,
+		CertData:   restConfig.TLSClientConfig.CertData,
+		KeyData:    restConfig.TLSClientConfig.KeyData,
+		ServerName: restConfig.TLSClientConfig.ServerName,
+	}
+	return nil
+}
+
+// NewRESTConfigForKubeConfig returns a REST Config for the given KubeConfigValue.
+func NewRESTConfigForKubeConfig(kubeConfig []byte) (*rest.Config, error) {
```
Comment (MisterMX) [inline pkg/controller/cluster/controller.go:561]: Should be lower case as its only used internally.

```suggestion
func NewRESTConfigForKubeConfig(kubeConfig []byte) (*rest.Config, error) {
```

File: pkg/controller/cluster/controller.go [hunk 8]
```diff
@@ -473,3 +516,52 @@ func (e *external) getPayload(ctx context.Context, ref *v1alpha1.SecretReference
 
 	return nil, nil
 }
+
+// ExtractKubeconfigFromSecretRef extracts login information from a Kubeconfig Secret Reference
+func (e *external) ExtractKubeconfigFromSecretRef(ctx context.Context, p *v1alpha1.ClusterParameters, r *argocdv1alpha1.Cluster) error {
```
Comment (MisterMX) [inline pkg/controller/cluster/controller.go:521]: Should be lower case as its only used internally.

```suggestion
func (e *external) extractKubeconfigFromSecretRef(ctx context.Context, p *v1alpha1.ClusterParameters, r *argocdv1alpha1.Cluster) error {
```

**Review by MisterMX — CHANGES_REQUESTED** [review]

**Comment by MisterMX** [comment]
Thanks for your PR @stevendborrelli! Can you take a look at the remarks and also solve the conflicts?

We would be happy to see this feature in `v0.2.0`. Do you think you can make it in the next couple of days?

**Comment by stevendborrelli** [comment]
@MisterMX I made an initial pass changing fields to optional but there are some issues in the Mock function causing tests to fail. I'll try to get something by Friday.

File: apis/cluster/v1alpha1/types.go [hunk 9]
```diff
@@ -59,13 +61,17 @@ type ClusterConfig struct {
...
 	// AWSAuthConfig contains IAM authentication configuration
 	// +optional
 	AWSAuthConfig *AWSAuthConfig `json:"awsAuthConfig,omitempty"`
 	// ExecProviderConfig contains configuration for an exec provider
 	// +optional
 	ExecProviderConfig *ExecProviderConfig `json:"execProviderConfig,omitempty"`
+	// KubeconfigSecretRef contains a reference to a kubernetes secret containing kubeconfig data
```
[the remark this answers is not in the document]
Comment (MisterMX) [reply apis/cluster/v1alpha1/types.go:72]:
```suggestion
	// KubeconfigSecretRef contains a reference to a Kubernetes secret entry that
	// contains a raw kubeconfig in YAML or JSON.
	// See https://kubernetes.io/docs/reference/config-api/kubeconfig.v1/ for more
	// info about Kubeconfigs.
```

[the remark this answers is not in the document]
Comment (stevendborrelli) [reply apis/cluster/v1alpha1/types.go:72]: Accepted

File: examples/cluster/cluster-kubeconfig.yaml (hunk 4 above)
Comment (stevendborrelli) [reply examples/cluster/cluster-kubeconfig.yaml:12]: accepted

File: pkg/controller/cluster/controller.go (hunk 6 above)
Comment (stevendborrelli) [reply pkg/controller/cluster/controller.go]: Removed

File: pkg/controller/cluster/controller.go (hunk 7 above)
Comment (stevendborrelli) [reply pkg/controller/cluster/controller.go]: renamed to lowercase

File: pkg/controller/cluster/controller.go (hunk 8 above)
Comment (stevendborrelli) [reply pkg/controller/cluster/controller.go]: renamed to lowercase

**Comment by deggja** [comment]
Any ETA on when this would be available for use? Sorry for asking :-)

**Comment by stevendborrelli** [comment]
Sorry for the delay everyone, I needed to add the ability to track changes to the kubeconfig, as cloud providers rotate the credentials. I've pushed a new commit that includes this tracking.

File: apis/cluster/v1alpha1/types.go (hunk 1 above)
Comment (stevendborrelli) [reply apis/cluster/v1alpha1/types.go]: I've updated the PR to make this field a pointer.

File: apis/cluster/v1alpha1/types.go (hunk 2 above)
Comment (stevendborrelli) [reply apis/cluster/v1alpha1/types.go]: Converted to a pointer.

**Comment by stevendborrelli** [comment]
The PR has been updated and is ready for review.

File: pkg/controller/cluster/controller.go [hunk 10]
```diff
@@ -120,16 +124,32 @@ func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.Ex
...
+		kubeconfigSecretResourceVersion, err := e.getResourceVersion(ctx, cr.Spec.ForProvider.Config.KubeconfigSecretRef)
+		if err != nil {
+			return managed.ExternalObservation{}, err
+		}
+		kubeconfigSecretUpToDate = cr.Status.AtProvider.Kubeconfig != nil &&
+			(kubeconfigSecretResourceVersion == currentStatus.Kubeconfig.Secret.ResourceVersion)
+		cr.Status.AtProvider.Kubeconfig = &v1alpha1.KubeconfigObservation{
+			Secret: v1alpha1.SecretObservation{ResourceVersion: kubeconfigSecretResourceVersion},
+		}
+	}
```
Comment (MisterMX) [inline pkg/controller/cluster/controller.go:145]: This should be in a separate function, i.e. `isKubeConfigSecretUpToDate`

File: pkg/controller/cluster/controller.go [hunk 11]
```diff
@@ -199,32 +220,36 @@ func lateInitializeCluster(p *v1alpha1.ClusterParameters, r *argocdv1alpha1.Clus
...
+		CacheInfo: &v1alpha1.ClusterCacheInfo{
+			ResourcesCount:    &r.Info.CacheInfo.ResourcesCount,
+			APIsCount:         &r.Info.CacheInfo.APIsCount,
+			LastCacheSyncTime: r.Info.CacheInfo.LastCacheSyncTime,
 		},
+		ApplicationsCount: r.Info.ApplicationsCount,
 	}
-
-	return o
+	return i
```
Comment (MisterMX) [inline pkg/controller/cluster/controller.go:248]: Is this rework necessary? I would prefer to have all status updates in a single function that is called from `Observe`.

**Review by MisterMX — CHANGES_REQUESTED** [review]
Thank you very much for your contribition @stevendborrelli. I added two remarks to your code. Can you take a look at them?

File: pkg/controller/cluster/controller.go (hunk 10 above)
Comment (stevendborrelli) [reply pkg/controller/cluster/controller.go]: I was able to compare using `cmp.Equal`

File: pkg/controller/cluster/controller.go (hunk 11 above)
Comment (stevendborrelli) [reply pkg/controller/cluster/controller.go]: I've reverted to the original return of `ClusterObservation` and added the update to the Kubeconfig Secret in this function.

**Comment by stevendborrelli** [comment]
@MisterMX I've squashed the changes for the PR.

**Review by MisterMX — APPROVED** [review]
LGTM. Thank you very much for your contribution @stevendborrelli!
